confgenerator/resourcedetector/gce_detector.go

```diff
@@ -0,0 +1,159 @@
+package resourcedetector
+
+import (
+	"fmt"
+)
+
+type gceAttribute int
+
+const (
+	project gceAttribute = iota
+	zone
+	network
+	subnetwork
+	publicIP
+	privateIP
+	instanceID
+	instanceName
+	tags
+	machineType
+	metadata
+	label
+	interfaceIPv4
+)
+
+func GetGCEDetector() (Detector, error) {
+	provider := NewGCEMetadataProvider()
+	dt := GCEDetectorBuilder{provider: provider}
+	return dt.GetDetector()
+}
+
+// The data provider interface for GCE environment
+// Implementation of this provider can use either the metadata server on VM,
+// or the cloud API
+type gceDataProvider interface {
+	getProject() (string, error)
+	getZone() (string, error)
+	getNetwork() (string, error)
+	getSubnetwork() (string, error)
+	getPublicIP() (string, error)
+	getPrivateIP() (string, error)
+	getInstanceID() (string, error)
+	getInstanceName() (string, error)
+	getTags() (string, error)
+	getMachineType() (string, error)
+	getLabelKeys() ([]string, error)
+	getLabel(key string) (string, error)
+	getMetadataKeys() ([]string, error)
+	getMetadata(key string) (string, error)
+	getInterfaceNames() ([]string, error)
+	getInterfaceIPv4(name string) (string, error)
+}
+
+// Nested attribute takes two steps: first to get the list of keys,
+// and then use the key to get individual values.
+type nestedAttribute struct {
```

Comment:
This struct and interface seems pretty confusing to me. It feels simpler to have the provider do both the fetching of the keys and then aggregate the values rather than having the detector work with this interface.

In other words, if the provider provided `getMetadata`, getLabels` and getLabels` methods that each returned a `map[string]string`, the GCE detector wouldn't need to do any of this nesting and unnesting logic right? That would be contained in the implementation of the `getMetadata`, getLabels` and getLabels` methods. Let me know what you think of this approach.